usb: device_next: Make stack High-Bandwidth aware

tmon-nordic opened the pull request:
Make stack aware of high-bandwidth endpoints and implement high-bandwidth functionality in DWC2 driver.

~~The first commit is reviewed at #74581.~~

tmon-nordic committed:
usb: Add missing wMaxPacketSize endian conversions

wMaxPacketSize in endpoint descriptor is stored in little-endian order,
but the mps parameter passed to functions is in host order.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
usb: device_next: Make stack High-Bandwidth aware

Add macros for converting between Max Packet Size and total payload
length. Allow drivers specify whether endpoint supports high-bandwidth
interrupt and high-bandwidth isochronous transfers.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
drivers: udc_dwc2: Allow larger TxFIFO writes

DMA is capable of packetizing transfers on its own and the driver
doesn't need to check available TxFIFO size when scheduling DMA
transfer.

Split packets on DWORD boundary when writing more than one packet to
TxFIFO in completer mode.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
drivers: udc_dwc2: Support High-Bandwidth endpoints

Make driver aware of High-Bandwidth endpoints both in Completer and
Buffer DMA mode. In Completer mode TxFIFO must be able to hold all
packets for microframe, while in Buffer DMA mode space enough for two
packets is sufficient.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
usb: device_next: hid: Simplify MPS less than 65 detection

Use COND_CODE_1() instead of UTIL_AND() to make it possible to only
define the HID_MPS_LESS_65_ macros up to value 64.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

jfischer-no commented on drivers/usb/udc/udc_common.c:
```diff
@@ -242,12 +242,16 @@ static bool ep_check_config(const struct device *dev,
 		break;
 	case USB_EP_TYPE_INTERRUPT:
-		if (!cfg->caps.interrupt) {
+		if ((!cfg->caps.interrupt) ||
```
`		if (!cfg->caps.interrupt ||`

jfischer-no commented on drivers/usb/udc/udc_common.c:
```diff
@@ -242,12 +242,16 @@ static bool ep_check_config(const struct device *dev,
 		break;
 	case USB_EP_TYPE_ISO:
-		if (!cfg->caps.iso) {
+		if ((!cfg->caps.iso) ||
```
`		if (!cfg->caps.iso ||`

jfischer-no commented on include/zephyr/usb/usb_ch9.h:
```diff
@@ -348,6 +348,29 @@ struct usb_association_descriptor {
+#define USB_MPS_ADDITIONAL_TRANSACTIONS(mps) (((mps) & 0x1800) >> 11)
+
+/** Calculate total payload length from Max Packet Size value */
+#define USB_MPS_TO_TPL(mps)	((1 + USB_MPS_ADDITIONAL_TRANSACTIONS(mps)) * \
+				 USB_MPS_EP_SIZE(mps))
```
```suggestion
#define USB_MPS_TO_TPL(mps)	 \
	 ((1 + USB_MPS_ADDITIONAL_TRANSACTIONS(mps)) * USB_MPS_EP_SIZE(mps))
```

jfischer-no commented on include/zephyr/drivers/usb/udc.h:
```diff
@@ -76,6 +76,10 @@ struct udc_ep_caps {
 	uint32_t bulk : 1;
 	/** ISO transfer capable endpoint */
 	uint32_t iso : 1;
+	/** High-Bandwidth interrupt transfer capable endpoint */
+	uint32_t hb_interrupt : 1;
+	/** High-Bandwidth isochronous transfer capable endpoint */
+	uint32_t hb_iso : 1;
```
I assume that if endpoint is capable of hb_interrupt, then it is capable of hb_iso. Is it not too cautious to have two flags?

tmon-nordic replied:
I wouldn't necessarily connect `hb_interrupt` with `hb_iso`. There are devices out there that support interrupt but do not support isochronous (and vice versa) on selected endpoints.

An alternative I was considering was to have `hb_capable` (or similar name) that would be a modifier. If `hb_capable` was set together with `interrupt` it would mean that endpoint can support both standard interrupt and high-bandwidth interrupt. Similarly `hb_capable` with `iso` would mean both standard iso and high-bandwidth iso.

I scrapped the modifier idea considering it is just a single bit in a bitfield... but if anything I would go with such modifier and not implying `hb_interrupt` means `hb_iso`.

tmon-nordic commented on subsys/usb/device_next/class/usbd_hid_macros.h:
```diff
@@ -84,975 +84,15 @@
-#define HID_ALL_MPS_LESS_65(n)							\
-	UTIL_AND(HID_MPS_LESS_65(DT_INST_PROP(n, out_report_size)),		\
+#define HID_ALL_MPS_LESS_65(n)									\
+	UTIL_AND(COND_CODE_1(HID_MPS_LESS_65(DT_INST_PROP_OR(n, out_report_size, 0)), (1), (0)),\
```
The `out_report_size` has to be read as `DT_INST_PROP_OR(n, out_report_size, 0)` because `out_report_size` is optional and the absence of it should trigger the "1" branch.

jfischer-no commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -46,11 +46,13 @@ K_MSGQ_DEFINE(drv_msgq, sizeof(struct dwc2_drv_event),
+ * Driver adds 2 locations for each OUT endpoint to this value.
  */
-#define UDC_DWC2_GRXFSIZ_DEFAULT	(15U + 512U/4U)
+#define UDC_DWC2_GRXFSIZ_DEFAULT	(13 + 1 + 774)
```
This will take up all available space on full-speed devices and they also do not support high bandwidth transfers, e.g. full-speed-only controllers on STM32F413 have only 2048 bytes of FIFO. I thought about different definitions for full/high speed and some Kconfig option, but it would be quite difficult to handle and not good to maintain. I think it is better to adjust it at runtime based on the configuration, e.g. in the udc_enable() implementation.

tmon-nordic replied:
For a proper fifo sizing, the driver must be aware of configurations and do allocations based on that. Because the RxFIFO size must be configured (thus must be known) before the configuration is selected, driver must have prior knowledge of the configurations to know what can be afforded. The configurations would also be necessary to determine whether the thresholding is necessary or not (it is recommended to not enable thresholding unless necessary).

For the time being, to not break STM32F413, as a temporary measure I can make it "dynamic" in `udc_dwc2_init_controller()`. The "dynamic" is because it would just be based on PHY speed and not based on the actual configuration (which is not exposed to UDC drivers).

jfischer-no committed:
drivers: udc: do not use the MPS value directly for the endpoint size

Instead, use the helper to get the size field from the MPS value.
MPS value may contain information about additional transaction
per microframe (bits 12..11).

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -853,8 +887,15 @@ static inline void dwc2_handle_rxflvl(const struct device *dev)
 
-		if (net_buf_tailroom(buf) && evt.bcnt == ep_cfg->mps) {
-			dwc2_prep_rx(dev, buf, ep_cfg, 0);
+		if (((evt.bcnt % USB_MPS_EP_SIZE(ep_cfg->mps)) == 0) &&
```
This PR does not change the use of the MPS value in the other drivers. We could consider ignoring it for the full speed controllers, but there are few drivers for full and high speed controllers, I think it is better to unify it and not use mps value directly. I also want to introduce udc_mps_ep_size() to get the size field from the endpoint configuration structure, see https://github.com/jfischer-no/zephyr/commit/1d84372d5d95a0b6864dc266afe84b5cb5be4c53#diff-e35b5db282dc4b82092896f84fea69d5be994a93c813eb07cad87f55b33c2523, [patch](https://github.com/jfischer-no/zephyr/commit/1d84372d5d95a0b6864dc266afe84b5cb5be4c53.patch)

tmon-nordic replied:
Other drivers do not set `caps.high_bandwidth` and therefore won't ever get mps that is greater than 1024. Only when the drivers do set `high_bandwidth` capability it is necessary to differentiate between wMaxPacketSize value and actual endpoint size.

I think the unification can come later.

jfischer-no replied:
I do not see why it should be later, you just have to cherrypick. This is the right time to unify it and also set a good example for new drivers.

tmon-nordic replied:
Cherry picked your commit and pushed. After the push, GitHub reported conflict in stm32 driver and for that reason I did rebase on top of main. Please review.

MarkWangChinese commented on drivers/usb/udc/udc_mcux_ehci.c:
```diff
@@ -571,7 +571,7 @@ static int udc_mcux_ep_enable(const struct device *dev,
 	ep_init.interval        = cfg->interval;
 	ep_init.endpointAddress = cfg->addr;
-	ep_init.maxPacketSize   = cfg->mps;
+	ep_init.maxPacketSize   = udc_mps_ep_size(cfg);
```
In NXP MCUX controller driver, the `maxPacketSize` contains the bits of additional transactions. So it is right to use `cfg->mps`?

tmon-nordic replied:
If `ep_init.maxPacketSize` is the same as `wMaxPacketSize` then it should be fine to use `cfg->mps` directly. If the driver correctly supports high-bandwidth endpoints, then yes `high_bandwidth` capability must be set (otherwise stack will fail to initialize if SW requests high-bandwidth endpoint).

MarkWangChinese replied:
HI @tmon-nordic , OK, Could you revert back the `maxPacketSize` changes of udc_mcux_ehci.c and udc_mdux_ip3511.c? I will create another pr to add the `high_bandwidth`. Is it OK?

tmon-nordic replied:
It was @jfischer-no requirement to cherry-pick his patch. I preferred to have this separate, because things are now piling up...

Can you just add the `high_bandwidth` (and remove the `udc_mps_ep_size()`) on top of this PR?

MarkWangChinese replied:
Added. please help to check.

tmon-nordic replied:
You need to fix the signed-off-by in the commit.

MarkWangChinese replied:
copy-paste issue. fixed.

jfischer-no replied:
@tmon-nordic (I beg you, because it is probably faster) Please add a comment above the lines `ep_init.maxPacketSize   = cfg->mps;`, otherwise no one else will know this internal of the NXP HAL drivers.

tmon-nordic replied:
Not sure if wording is what you expected, but I went in with "HAL expects wMaxPacketSize value directly in maxPacketSize field".

MarkWangChinese committed:
drivers: udc: mcux: set high_bandwidth

set high_bandwidth and use the cfg->mps to set
ep_init.maxPacketSize because NXP MCUX controller
driver supports additional transaction bits in
maxPacketSize.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>
